File: packages/validation/pipeline.py
"""Validation pipeline types and orchestration.

Defines the core types (Severity, ValidationIssue, ValidationResult) and the
ValidationPipeline that chains multiple validators in sequence.
"""

from __future__ import annotations

from dataclasses import dataclass
from enum import StrEnum
from typing import TYPE_CHECKING, Any, Protocol

if TYPE_CHECKING:
    from collections.abc import Sequence


class Severity(StrEnum):
    """Validation issue severity levels."""

    ERROR = "error"
    WARNING = "warning"
    INFO = "info"


@dataclass(frozen=True, slots=True)
class ValidationIssue:
    """Single validation issue with severity, message, and optional location."""

    severity: Severity
    message: str
    location: str = ""


@dataclass(frozen=True, slots=True)
class ValidationResult:
    """Aggregated result of running one or more validators."""

    issues: tuple[ValidationIssue, ...]

    @property
    def is_valid(self) -> bool:
        """True when no ERROR-level issues exist."""
        return not any(i.severity == Severity.ERROR for i in self.issues)

    def errors(self) -> tuple[ValidationIssue, ...]:
        """Return only ERROR-level issues."""
        return tuple(i for i in self.issues if i.severity == Severity.ERROR)

    def warnings(self) -> tuple[ValidationIssue, ...]:
        """Return only WARNING-level issues."""
        return tuple(i for i in self.issues if i.severity == Severity.WARNING)

    @staticmethod
    def merge(*results: ValidationResult) -> ValidationResult:
        """Combine multiple results into one."""
        all_issues: list[ValidationIssue] = []
        for r in results:
            all_issues.extend(r.issues)
        return ValidationResult(issues=tuple(all_issues))

    @staticmethod
    def ok() -> ValidationResult:
        """Create an empty (passing) result."""
        return ValidationResult(issues=())


class Validator(Protocol):
    """Protocol that all validators must satisfy."""

    def validate(
        self,
        *,
        front: str,
        back: str,
        tags: tuple[str, ...] = (),
        **kwargs: Any,
    ) -> ValidationResult: ...


class ValidationPipeline:
    """Runs a sequence of validators and aggregates their results."""

    def __init__(self, validators: Sequence[Validator]) -> None:
        self._validators = tuple(validators)

    def run(
        self,
        *,
        front: str,
        back: str,
        tags: tuple[str, ...] = (),
        **kwargs: Any,
    ) -> ValidationResult:
        """Execute all validators and merge results."""
        if not self._validators:
            return ValidationResult.ok()
        results = [
            v.validate(front=front, back=back, tags=tags, **kwargs) for v in self._validators
        ]
        return ValidationResult.merge(*results)

File: crates/validation/src/pipeline.rs
use serde::{Deserialize, Serialize};

/// Severity of a validation issue.
#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum Severity {
    Error,
    Warning,
    Info,
}

impl std::fmt::Display for Severity {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            Severity::Error => write!(f, "error"),
            Severity::Warning => write!(f, "warning"),
            Severity::Info => write!(f, "info"),
        }
    }
}

/// A single validation issue.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct ValidationIssue {
    pub severity: Severity,
    pub message: String,
    /// Optional location hint (e.g. "front", "back", "tags").
    pub location: String,
}

impl ValidationIssue {
    pub fn error(message: impl Into<String>, location: impl Into<String>) -> Self {
        Self {
            severity: Severity::Error,
            message: message.into(),
            location: location.into(),
        }
    }

    pub fn warning(message: impl Into<String>, location: impl Into<String>) -> Self {
        Self {
            severity: Severity::Warning,
            message: message.into(),
            location: location.into(),
        }
    }

    pub fn info(message: impl Into<String>, location: impl Into<String>) -> Self {
        Self {
            severity: Severity::Info,
            message: message.into(),
            location: location.into(),
        }
    }
}

/// Aggregated validation result.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct ValidationResult {
    pub issues: Vec<ValidationIssue>,
}

impl ValidationResult {
    /// True when no ERROR-level issues exist.
    pub fn is_valid(&self) -> bool {
        !self.issues.iter().any(|i| i.severity == Severity::Error)
    }

    /// Filter to ERROR-level issues only.
    pub fn errors(&self) -> Vec<&ValidationIssue> {
        self.issues
            .iter()
            .filter(|i| i.severity == Severity::Error)
            .collect()
    }

    /// Filter to WARNING-level issues only.
    pub fn warnings(&self) -> Vec<&ValidationIssue> {
        self.issues
            .iter()
            .filter(|i| i.severity == Severity::Warning)
            .collect()
    }

    /// Merge multiple results into one.
    pub fn merge(results: &[ValidationResult]) -> ValidationResult {
        ValidationResult {
            issues: results
                .iter()
                .flat_map(|r| r.issues.clone())
                .collect(),
        }
    }

    /// Empty (passing) result.
    pub fn ok() -> ValidationResult {
        ValidationResult {
            issues: Vec::new(),
        }
    }
}

/// Trait for validators. All implementations must be Send + Sync.
pub trait Validator: Send + Sync {
    fn validate(&self, front: &str, back: &str, tags: &[String]) -> ValidationResult;
}

/// Pipeline that chains validators in sequence and merges results.
pub struct ValidationPipeline {
    validators: Vec<Box<dyn Validator>>,
}

impl ValidationPipeline {
    pub fn new(validators: Vec<Box<dyn Validator>>) -> Self {
        Self { validators }
    }

    /// Run all validators and return merged result.
    pub fn run(&self, front: &str, back: &str, tags: &[String]) -> ValidationResult {
        let results: Vec<ValidationResult> = self
            .validators
            .iter()
            .map(|v| v.validate(front, back, tags))
            .collect();
        ValidationResult::merge(&results)
    }
}
